lambda/index.js

```diff
@@ -1,92 +1,74 @@
-                    const params = {
-                        MessageBody: JSON.stringify({Message: JSON.stringify(message)}),
-                        QueueUrl: queueUrl,
-                        DelaySeconds: RETRY_TIMOUT
-                    };
-                    const msg = `[RETRY] retrying ${invalidationParams.DistributionId}:${
-                        invalidationParams.Paths
-                        }`;
-                    console.log(params);
-
-                    sqs.sendMessage(params, (err, data) => {
-                        if (err) {
-                            console.log(err);
-                            callback("failed to send message for retry");
-                            return;
-                        }
-                        const msg = `[RETRY] Scheduling a retry in ${RETRY_TIMOUT} seconds. ${retried} retries`;
-                        console.log(msg);
-                        callback(null, msg);
-                        return;
-
-                    });
-                }
-            } else
-                console.log(data);
-
-        });
+const NUM_OF_RETRIES = process.env.INVALIDATION_MAX_RETRIES;
 
-    });
+const handleInvalidationError = async err => {
+  if (err.code !== "TooManyInvalidationsInProgress") {
+    const warning = `ignoring error: ${err}.`;
+    logWarning(warning);
+    return Promise.resolve();
+  }
+
+  if (shouldRetry(message)) {
+    try {
+      return await retryMessage(message);
+    } catch (err) {
+      log(err);
+      return Promise.resolve("failed to send message for retry");
+    }
+  } else {
+    const warning = `Failed after ${NUM_OF_RETRIES} retries`;
+    logWarning(warning);
+    return Promise.resolve(warning);
+  }
+};
 
-    callback(null, "A-OK");
+async function retryMessage(message) {
+  message.retry_count++;
+  // This Message must be in SNS => SQS Format for incoming parsing to work
+  const params = makeSnsParams(message, record.eventSourceARN);
+  logRetry(
+    `retrying ${invalidationParams.DistributionId}:${invalidationParams.Paths}`,
+    params
+  );
+  return await sendSqsMessage(params);
+}
+
+function shouldRetry(message) {
+  let retried = message.retry_count || 0;
+  return retried < NUM_OF_RETRIES - 1;
+}
+
+exports.handler = async function(event) {
+  await asyncForEach(event.Records, async record => {
+    const message = parseMessage(record);
+    log("SQS Message: ", message);
+
+    if (!message.distribution_id || !message.path) {
+      logWarning(
+        'bad format. desired SNS message format: {"distribution_id": "<distid>", "path": "/a/path/*"}'
+      );
+      return Promise.resolve();
+    }
+
+    try {
+      const successfulInvalidation = await createCloudfrontInvalidation(
+        makeInvalidationParams(message)
+      );
+      log(successfulInvalidation);
+      return Promise.resolve();
+    } catch (e) {
+      return await handleInvalidationError(e);
```

Comment:
Should either supply message or event or record (I'm not an expert on JS, so sort of guessing which is appropriate) and have handleInvalidationError handle it appropriately.